guides/source/asset_pipeline.md

```diff
@@ -8,479 +8,618 @@ This guide covers the asset pipeline.
+    tools like [Webpack](https://webpack.js.org/),
+    [esbuild](https://esbuild.github.io/), or [Vite](https://vite.dev/) into
+    your Rails application, enabling you to handle complex asset requirements
+    alongside Propshaft.
+
+    Use `js-bundling-rails` with these tools to bundle JavaScript files, manage
+    dependencies, and transpile modern JavaScript for browser compatibility. For
+    more details, check the [`js-bundling-rails`
+    documentation](https://github.com/rails/jsbundling-rails).
+
+    Use `css-bundling-rails` to pre-process stylesheets using tools like
+    [Sass](https://sass-lang.com/) or [Tailwind CSS](https://tailwindcss.com/),
+    making it easier to work with advanced CSS features. For more details, check
+    the [`css-bundling-rails`
+    documentation](https://github.com/rails/cssbundling-rails).
+
+    To read more about how to use these gems you can see the [Advanced Asset
+    Management section](#advanced-asset-management). These tools complement
+    Propshaft by handling the complex processing, while Propshaft efficiently
+    organizes and serves the final assets.
+
+### Versioning with digest-based URLs
+
+In Rails, asset versioning helps manage cache behavior by adding unique
+identifiers to asset filenames. This ensures that whenever an asset is updated,
+the browser sees it as a new file and loads the latest version instead of using
+an older, cached one.
+
+Sprockets automatically handles asset fingerprinting by appending a hash to
+filenames whenever assets are updated. Similarly, Propshaft uses content-based
+fingerprinting and automatically updates asset paths in CSS to reference
+digested filenames. However, for JavaScript files, you’ll need to manually
+trigger the transformation or use a bundler to ensure the correct filenames are
+applied.
+
+You can read more about how to use asset versioning in the [Fingerprinting
+section](#fingerprinting).
+
+### Asset Precompilation
+
+In Propshaft, all assets from the paths configured in `config.assets.paths` are
+available for serving and will be copied into `public/assets` when precompiling.
+Unlike Sprockets, which only copied assets that were explicitly included in a
+bundle, Propshaft automatically processes all assets from the specified paths.
+This includes images, stylesheets, JavaScript files, and more.
+
+If you want to exclude certain directories from this process, you can use
+`config.assets.excluded_paths`. Read more about this in the [Fingerprinting
+section](#fingerprinting).
+
+Once precompiled, these assets can be referenced by their logical path using
+helpers like `asset_path`, `image_tag`, and `javascript_include_tag`. In production,
+when `assets:precompile` is run, these references are automatically turned into
+digest-aware paths. This is done using a `manifest.json` file that is
+automatically generated in` public/assets/manifest.json`. This file maps logical
+asset paths to their precompiled file paths.
```

Comment:
Yup! Thanks for calling this out. This and a few other sections were repeated because of my structure, however, I’ve re-ordered the sections so that the flow feels much more natural, and it helped me catch and remove some duplications.